src/internal/task/etcd_task_service.go

```diff
@@ -0,0 +1,340 @@
+package task
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	col "github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachhash"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+)
+
+const (
+	groupPrefix = "/group"
+	taskPrefix  = "/task"
+	claimPrefix = "/claim"
+)
+
+type Service interface {
+	// TODO: Should group ID parameter?
+	Doer(ctx context.Context, namespace, groupID string, cb func(context.Context, Doer) error) error
+	Maker(namespace string) Maker
+}
+
+type etcdService struct {
+	etcdClient *etcd.Client
+	etcdPrefix string
+}
+
+func NewEtcdService(etcdClient *etcd.Client, etcdPrefix string) Service {
+	return &etcdService{
+		etcdClient: etcdClient,
+		etcdPrefix: etcdPrefix,
+	}
+}
+
+func (es *etcdService) Doer(ctx context.Context, namespace, groupID string, cb func(context.Context, Doer) error) error {
+	namespaceEtcd := newNamespaceEtcd(es.etcdClient, es.etcdPrefix, namespace)
+	return namespaceEtcd.groupCol.WithRenewer(ctx, func(ctx context.Context, renewer *col.Renewer) error {
+		if err := renewer.Put(ctx, groupID, &Group{}); err != nil {
+			return err
+		}
+		return namespaceEtcd.taskCol.WithRenewer(ctx, func(ctx context.Context, renewer *col.Renewer) error {
+			ed := newEtcdDoer(namespaceEtcd, groupID, renewer)
+			return cb(ctx, ed)
+		})
+	})
+}
+
+type namespaceEtcd struct {
+	etcdClient                  *etcd.Client
+	groupCol, taskCol, claimCol col.EtcdCollection
+}
+
+func newNamespaceEtcd(etcdClient *etcd.Client, etcdPrefix, namespace string) *namespaceEtcd {
+	return &namespaceEtcd{
+		etcdClient: etcdClient,
+		groupCol:   newCollection(etcdClient, path.Join(etcdPrefix, groupPrefix, namespace), &Group{}),
+		taskCol:    newCollection(etcdClient, path.Join(etcdPrefix, taskPrefix, namespace), &Task{}),
+		claimCol:   newCollection(etcdClient, path.Join(etcdPrefix, claimPrefix, namespace), &Claim{}),
+	}
+}
+
+func newCollection(etcdClient *etcd.Client, etcdPrefix string, template proto.Message) col.EtcdCollection {
+	return col.NewEtcdCollection(
+		etcdClient,
+		etcdPrefix,
+		nil,
+		template,
+		nil,
+		nil,
+	)
+}
+
+type Doer interface {
```

Comment:
I think the `Doer` and `Maker` interfaces should be with the Service so the whole model is contiguous in one place.  We also need to comments on those interfaces.